mlc_llm/relax_model/commons.py

```diff
@@ -70,50 +80,157 @@ def shard_gate_up_weight_scale(weight: relax.TensorStructInfo):
     for _, param in param_manager.params.items():
         if param.shard_strategy is None:
             pass
-        elif param.shard_strategy == "shard_qkv":
-            for i, weight in enumerate(param_manager.param2qrange[param]):
-                name = f"shard_qkv_{i}"
-                if name not in shard_funcs:
-                    shard_funcs[name] = shard_qkv_weight_scale(q_params[weight])
-                add_to_shard_info(f"param_{weight}", name)
-        elif param.shard_strategy == "shard_mlp_k":
-            for i, weight in enumerate(param_manager.param2qrange[param]):
-                name = f"shard_mlp_k_{i}"
-                if name not in shard_funcs:
-                    shard_funcs[name] = shard_k_weight_scale(q_params[weight])
-                add_to_shard_info(f"param_{weight}", name)
-        elif param.shard_strategy == "shard_o_proj_k":
-            for i, weight in enumerate(param_manager.param2qrange[param]):
-                name = f"shard_o_proj_k_{i}"
-                if name not in shard_funcs:
-                    shard_funcs[name] = shard_k_weight_scale(q_params[weight])
-                add_to_shard_info(f"param_{weight}", name)
-        elif param.shard_strategy == "shard_gate_up":
+        elif param.shard_strategy in shard_strategy_to_func:
             for i, weight in enumerate(param_manager.param2qrange[param]):
-                name = f"shard_gate_up_{i}"
-                if name not in shard_funcs:
-                    shard_funcs[name] = shard_gate_up_weight_scale(q_params[weight])
-                add_to_shard_info(f"param_{weight}", name)
+                if args.use_presharded_weights:
+                    sharding_func_name = None
+                else:
+                    sharding_func_name = f"{param.shard_strategy}_{i}"
+                    if sharding_func_name not in shard_funcs:
+                        shard_funcs[sharding_func_name] = shard_strategy_to_func[
+                            param.shard_strategy
+                        ](q_params[weight])
+                add_to_shard_info(f"param_{weight}", sharding_func_name)
         else:
             raise NotImplementedError(f"Shard strategy not implemented: {param.shard_strategy}")
+
+    bb = relax.BlockBuilder()  # pylint: disable=invalid-name
+
     for name, func in shard_funcs.items():
         func = func.with_attr({"global_symbol": name})
-        mod[name] = func
-    bb = relax.BlockBuilder()  # pylint: disable=invalid-name
+        bb.add_func(func, name)
+
     with bb.function("get_shard_info", params=[]):
         bb.emit_func_output(relax.StringImm(json.dumps(shard_info_dict)))
-    mod["get_shard_info"] = bb.get()["get_shard_info"]
+
+    return bb.get()
+
+
+def create_shard_transformation_func(param_manager, args, model_config) -> tvm.IRModule:
+    shard_strategy_to_func = _get_shard_strategies(
+        model_config,
+        num_shards=args.num_shards,
+        param_shape_is_already_sharded=args.build_model_only,
```

Comment:
Why does `build_model_only` mean `param_shape_is_already_sharded`?

Reply:
Currently, the initial model definition produces sharded shapes for `--build-model-only`, and unsharded shapes for `--convert-weight-only`.  (Based on [this conditional](https://github.com/mlc-ai/mlc-llm/blob/main/mlc_llm/relax_model/llama.py#L63).)  Ideally, this would be part of a later transformation applied to the the model definition, rather than part of the original model definition, and that's something I want to improve in the future.

For now, since `_get_shard_strategies` may be called either with `--build-model-only` (un-sharded weights, produce sharding functions in order to save alongside model) or with `--convert-weight-only` (pre-sharded weights, produce sharding functions in order to apply at compile time), it needs to know whether the parameter shapes in the model have been sharded or not.

Reply:
Thanks it makes sense, I didn't know that `num-shards` is ignored in the model def when `--convert-weight-only`.